server/global/data_analyzer/analyzer.js

```diff
@@ -0,0 +1,298 @@
+ * @param {object[]} molecules The molecule list
+ * @returns {{molecule : string, property : string, value : string}[]}
+ */
+function getNonValidNumberValue(molecules) {
+  return molecules.reduce(
+    (noNumberValues, molecule) =>
+      ["level_easy", "level_hard", "ntr"].reduce((noNumberValues, prop) => {
+        const value = molecule[prop];
+        if (value !== null && (!isNumber(value) || value < 0 || value > 1)) {
+          noNumberValues.push({ molecule: molecule.dci, property: prop, value });
+        }
+        return noNumberValues;
+      }, noNumberValues),
+    []
+  );
+}
+
+export class AnalyzerWarning {
+  constructor(type, message) {
+    this.type = type;
+    this.message = message;
+  }
+}
+
+AnalyzerWarning.DUPLICATE_UNIQUE_VALUE = 1;
+AnalyzerWarning.TOO_LONG_VALUE = 2;
+AnalyzerWarning.TOO_CLOSE_VALUES = 3;
+AnalyzerWarning.DUPLICATE_CLASSIFICATION_NODE = 4;
+AnalyzerWarning.INVALID_TYPE = 5;
+
+/**
+ * Returns values that appear more than once in the list
+ * @param {any[]} values
+ * @returns {any[]} non-unique values
+ */
+function getDuplicates(values) {
+  return [...new Set(values.filter((value, i) => values.slice(i + 1).includes(value)))];
+}
+
+/**
+ * Get all values which are longer that the max length
+ * @param {string[]} values
+ * @param {number} maxlength
+ * @returns {string[]} The list a too long values
+ */
+function getTooLongValues(values, maxlength) {
+  return values.filter((value) => value.length > maxlength);
+}
+
+/**
+ * Returns all groups of values that have a distance less than a given one
+ * @param {string[]} values The values
+ * @param {number} minDistance The maximum distance
+ * @returns {string[][]}
+ */
+function getTooCloseValues(values, minDistance) {
+  values = values.slice();
+  const groups = [];
+
+  values.forEach((value) => {
+    if (!isString(value)) {
+      return;
+    }
+    const group = values.filter((other) => {
+      if (!isString(other)) {
+        return;
+      }
+      const distance = levenshtein(other, value);
+      return distance <= minDistance && distance > 0;
+    });
+
+    // console.log(value, group);
+
+    const existingGroup = groups.find((egroup) => egroup.some((e) => group.includes(e)));
+    if (existingGroup) {
+      existingGroup.push(...group, value);
+      return;
+    }
+
+    if (group.length > 0) {
+      groups.push([...group, value]);
+    }
+  });
+
+  return groups.map((group) => [...new Set(group)]);
+}
+
+/**
+ * Find nodes with the same name but a different parent
+ * @param {object[]} classification The set of nodes
+ * @returns {{node : string, parents : string[]}[]}
+ */
+function findNodeWithDifferentsParents(classification) {
+  classification = flattenClassification(classification);
+  const groups = [];
+  classification.forEach((node, i) => {
+    if (!node) {
+      return;
+    }
+
+    const duplicates = classification
+      .slice(i)
+      .filter((other) => other && node.name === other.name)
```

Comment:
All the names in the filtered array are equals to `node.name`, so is it useful to keep all of them in the array instead of a simple length, without the call to `map()`?
Later we can use the value of `node.name` instead of `duplicates[0]` in the comparison.
(unless I misunderstood something, which is totally possible)

Reply:
You're totally right ! I changed that in [my last commit](https://github.com/ValFraNath/glowing-octo-guacamole/pull/174/commits/181ca3da96ea1037c1445ec4717642c8740ad7c7).